nixos/modules/services/networking/dnscrypt-proxy2.nix

```diff
@@ -35,7 +44,25 @@ in
       example = "/etc/dnscrypt-proxy/dnscrypt-proxy.toml";
       type = types.path;
       default = pkgs.runCommand "dnscrypt-proxy.toml" {
-        json = builtins.toJSON cfg.settings;
+        json =
+          let
+            # Settings extracted from the upstream example
+            exampleSettings = builtins.fromJSON (
+              builtins.readFile (
+                pkgs.runCommand "example-dnscrypt-proxy.json" {} ''
+                  ${pkgs.remarshal}/bin/toml2json ${pkgs.dnscrypt-proxy2.src}/dnscrypt-proxy/example-dnscrypt-proxy.toml > $out
+                ''
+              )
```

Comment:
this looks like IFD, and I think Hydra will fail to evaluate it while building the list of options with default values.

Reply:
I'm not entirely sure what exactly that means but I think if I moved the configFile generation to `config =` instead, that wouldn't be an issue right?

Reply:
It means that to evaluate (compute the default value of the option)
nix needs to build a derivation (here pkgs.dnscrypt-proxy2 and
pkgs.remarshal). This is forbidden on Hydra.

I don't really understand what you mean by "moving configFile genration
to `config =`".

Reply:
I see. The remarshal derivation has been in the default value since this module's inception though, so I thought it was fine.

I meant moving it out of the options.services.dnscrypt-proxy2 attrset and into the config one. The config attrset shouldn't be evaluated by hydra, right?

Reply:
Not by default but it might, if we try to create tests...

Reply:
How big is the example configuration? Could we just copy over the defaults?

Reply:
There is a test already: https://github.com/NixOS/nixpkgs/blob/74e4cb7ea4a32cecf70bb92f37fea66f3bcc1f15/nixos/tests/dnscrypt-proxy2.nix

The toml generated from the example config is 52 significant lines and has 5 keys/sections.  
The example config with comments etc. is 686 lines and gets updated every now and then when new features arrive https://github.com/DNSCrypt/dnscrypt-proxy/blob/master/dnscrypt-proxy/example-dnscrypt-proxy.toml

Remarshal is needed at least once anyways since there's no way to convert the settings attrset to TOML directly in Nix afaik.

Reply:
The alternative would be to merge both in a derivation without importing it with external tools.

Reply:
Do you mean that we should merge the TOML files directly instead of updating attrsets?

Reply:
Yes. I suppose one could write a python script that deep merges the content of both configuration files.

Reply:
We'd still need to run remarshal to convert the settings attrset to TOML though. Or would that be fine?

Reply:
Yes. If you don't import the configuration into nix, this is not import from derivation.

Reply:
I see. The `builtins.fromJSON` is the problematic part then, right?

Reply:
Yes. You need to do the merge in a derivation without nix. i.e. `runCommand`.

Reply:
What I just pushed should not require IFD anymore.

Reply:
Yes it looks better :)